File: adapters/executor/claude.go
package executor

import (
	"context"
	"encoding/json"
	"fmt"
	"strings"

	"github.com/anthropics/anthropic-sdk-go"
	"github.com/janmarkuslanger/stagehand/ports"
)

const maxAgentSteps = 20
const defaultModel = "claude-opus-4-7"

// ClaudeExecutor runs a task by calling the Anthropic Messages API and
// dispatching tool calls to the provided ArtifactStorage.
type ClaudeExecutor struct {
	client  *anthropic.Client
	storage ports.ArtifactStorage
}

// NewClaudeExecutor creates a ClaudeExecutor backed by the given API client and storage.
func NewClaudeExecutor(client *anthropic.Client, storage ports.ArtifactStorage) *ClaudeExecutor {
	return &ClaudeExecutor{client: client, storage: storage}
}

// Execute runs the task described by request, returning the agent's final text
// output and the paths of any files written during execution.
func (e *ClaudeExecutor) Execute(ctx context.Context, request ports.ExecutionRequest) (ports.ExecutionResult, error) {
	model := request.Model
	if model == "" {
		model = defaultModel
	}

	systemPrompt := request.SystemPrompt
	if systemPrompt == "" {
		systemPrompt = "You are a helpful AI assistant."
	}

	tools := buildTools(request.Tools)

	messages := []anthropic.MessageParam{
		anthropic.NewUserMessage(anthropic.NewTextBlock(request.Prompt)),
	}

	var finalOutput string
	var writtenFiles []string
	var lastStopReason anthropic.StopReason

	for step := 0; step < maxAgentSteps; step++ {
		resp, err := e.client.Messages.New(ctx, anthropic.MessageNewParams{
			Model:     anthropic.Model(model),
			MaxTokens: 16000,
			System: []anthropic.TextBlockParam{
				{
					Text:         systemPrompt,
					CacheControl: anthropic.NewCacheControlEphemeralParam(),
				},
			},
			Messages: messages,
			Tools:    tools,
		})
		if err != nil {
			return ports.ExecutionResult{}, fmt.Errorf("claude executor: task %s: step %d: %w", request.TaskID, step, err)
		}

		lastStopReason = resp.StopReason

		for _, block := range resp.Content {
			if textBlock, ok := block.AsAny().(anthropic.TextBlock); ok {
				finalOutput = textBlock.Text
			}
		}

		if resp.StopReason != anthropic.StopReasonToolUse {
			break
		}

		messages = append(messages, resp.ToParam())

		var toolResults []anthropic.ContentBlockParamUnion
		for _, block := range resp.Content {
			toolUse, ok := block.AsAny().(anthropic.ToolUseBlock)
			if !ok {
				continue
			}
			resultContent, toolErr := e.dispatchTool(ctx, request.TaskID, toolUse, &writtenFiles)
			if toolErr != nil {
				toolResults = append(toolResults, anthropic.NewToolResultBlock(toolUse.ID, toolErr.Error(), true))
			} else {
				toolResults = append(toolResults, anthropic.NewToolResultBlock(toolUse.ID, resultContent, false))
			}
		}

		messages = append(messages, anthropic.NewUserMessage(toolResults...))
	}

	if lastStopReason == anthropic.StopReasonToolUse {
		return ports.ExecutionResult{}, fmt.Errorf("claude executor: task %s: agent did not complete within %d steps", request.TaskID, maxAgentSteps)
	}

	return ports.ExecutionResult{Output: finalOutput, Files: writtenFiles}, nil
}

func (e *ClaudeExecutor) dispatchTool(ctx context.Context, taskID string, toolUse anthropic.ToolUseBlock, writtenFiles *[]string) (string, error) {
	switch toolUse.Name {
	case "write_file":
		return e.executeWriteFile(ctx, taskID, toolUse.Input, writtenFiles)
	case "read_file":
		return e.executeReadFile(ctx, taskID, toolUse.Input)
	case "list_files":
		return e.executeListFiles(ctx, taskID, toolUse.Input)
	default:
		return "", fmt.Errorf("unknown tool %q", toolUse.Name)
	}
}

type writeFileInput struct {
	Path    string `json:"path"`
	Content string `json:"content"`
}

func (e *ClaudeExecutor) executeWriteFile(ctx context.Context, taskID string, rawInput json.RawMessage, writtenFiles *[]string) (string, error) {
	var input writeFileInput
	if err := json.Unmarshal(rawInput, &input); err != nil {
		return "", fmt.Errorf("write_file: invalid input: %w", err)
	}
	if input.Path == "" {
		return "", fmt.Errorf("write_file: path is required")
	}
	storagePath := taskID + "/" + input.Path
	if err := e.storage.Write(ctx, storagePath, []byte(input.Content)); err != nil {
		return "", fmt.Errorf("write_file: %w", err)
	}
	*writtenFiles = append(*writtenFiles, storagePath)
	return "ok", nil
}

type readFileInput struct {
	Path string `json:"path"`
}

func (e *ClaudeExecutor) executeReadFile(ctx context.Context, taskID string, rawInput json.RawMessage) (string, error) {
	var input readFileInput
	if err := json.Unmarshal(rawInput, &input); err != nil {
		return "", fmt.Errorf("read_file: invalid input: %w", err)
	}
	if input.Path == "" {
		return "", fmt.Errorf("read_file: path is required")
	}
	data, err := e.storage.Read(ctx, taskID+"/"+input.Path)
	if err != nil {
		return "", fmt.Errorf("read_file: %w", err)
	}
	return string(data), nil
}

type listFilesInput struct {
	Pattern string `json:"pattern"`
}

func (e *ClaudeExecutor) executeListFiles(ctx context.Context, taskID string, rawInput json.RawMessage) (string, error) {
	var input listFilesInput
	if err := json.Unmarshal(rawInput, &input); err != nil {
		return "", fmt.Errorf("list_files: invalid input: %w", err)
	}
	pattern := taskID + "/*"
	if input.Pattern != "" {
		pattern = taskID + "/" + input.Pattern
	}
	files, err := e.storage.List(ctx, pattern)
	if err != nil {
		return "", fmt.Errorf("list_files: %w", err)
	}
	return strings.Join(files, "\n"), nil
}

func buildTools(toolNames []string) []anthropic.ToolUnionParam {
	available := map[string]anthropic.ToolParam{
		"write_file": {
			Name:        "write_file",
			Description: anthropic.String("Write text content to a file. Creates the file if it does not exist."),
			InputSchema: anthropic.ToolInputSchemaParam{
				Properties: map[string]any{
					"path":    map[string]any{"type": "string", "description": "Relative file path"},
					"content": map[string]any{"type": "string", "description": "Text content to write"},
				},
				Required: []string{"path", "content"},
			},
		},
		"read_file": {
			Name:        "read_file",
			Description: anthropic.String("Read the text content of a file."),
			InputSchema: anthropic.ToolInputSchemaParam{
				Properties: map[string]any{
					"path": map[string]any{"type": "string", "description": "Relative file path"},
				},
				Required: []string{"path"},
			},
		},
		"list_files": {
			Name:        "list_files",
			Description: anthropic.String("List files matching a glob pattern."),
			InputSchema: anthropic.ToolInputSchemaParam{
				Properties: map[string]any{
					"pattern": map[string]any{"type": "string", "description": "Glob pattern, e.g. *.md"},
				},
				Required: []string{"pattern"},
			},
		},
	}

	var result []anthropic.ToolUnionParam
	for _, name := range toolNames {
		if tool, ok := available[name]; ok {
			toolCopy := tool
			result = append(result, anthropic.ToolUnionParam{OfTool: &toolCopy})
		}
	}
	return result
}

// Compile-time check that ClaudeExecutor satisfies ports.AgentExecutor.
var _ ports.AgentExecutor = (*ClaudeExecutor)(nil)

File: stagehand/adapters/executor/claude.py
from __future__ import annotations

import asyncio
from typing import Any, Optional

import anthropic

from stagehand.adapters.executor.base import (
    BaseAgentExecutor,
    ParsedTurn,
    ToolInvocation,
)
from stagehand.adapters.logger import NullLogger
from stagehand.ports.executor import ExecutionRequest, ToolDefinition
from stagehand.ports.logger import Logger
from stagehand.ports.storage import ArtifactStorage

DEFAULT_MODEL = "claude-opus-4-5"


class ClaudeExecutor(BaseAgentExecutor):
    """Runs a task by calling the Anthropic Messages API.

    Extra tools can be passed at construction time, enabling extension
    without modifying core code:

        executor = ClaudeExecutor(api_key="...", extra_tools=[MyTool])
    """

    _label = "claude executor"

    def __init__(
        self,
        api_key: Optional[str] = None,
        storage: Optional[ArtifactStorage] = None,
        extra_tools: Optional[list[ToolDefinition]] = None,
        rate_limit_retries: int = 3,
        rate_limit_delay: float = 60.0,
        logger: Optional[Logger] = None,
    ) -> None:
        super().__init__(storage=storage, extra_tools=extra_tools)
        self.client = anthropic.AsyncAnthropic(api_key=api_key)
        self.rate_limit_retries = rate_limit_retries
        self.rate_limit_delay = rate_limit_delay
        self._logger: Logger = logger or NullLogger()

    def _default_model(self) -> str:
        return DEFAULT_MODEL

    def _init_messages(self, request: ExecutionRequest) -> list[Any]:
        return [{"role": "user", "content": request.prompt}]

    def _serialize_tools(self, tools: list[ToolDefinition]) -> list[anthropic.types.ToolParam]:
        return [
            anthropic.types.ToolParam(
                name=t.name,
                description=t.description,
                input_schema=t.input_schema,
            )
            for t in tools
        ]

    async def _call_model(
        self,
        model: str,
        request: ExecutionRequest,
        messages: list[Any],
        tools: list[anthropic.types.ToolParam],
    ) -> anthropic.types.Message:
        system_prompt = request.system_prompt or "You are a helpful AI assistant."
        return await self._create_with_retry(model, system_prompt, messages, tools)

    def _parse_response(
        self, resp: anthropic.types.Message, request: ExecutionRequest, step: int
    ) -> ParsedTurn:
        text: Optional[str] = None
        tool_calls: list[ToolInvocation] = []
        for block in resp.content:
            if block.type == "text":
                text = block.text
            elif block.type == "tool_use":
                tool_calls.append(
                    ToolInvocation(id=block.id, name=block.name, raw_arguments=block.input)
                )

        is_complete = resp.stop_reason != "tool_use"
        assistant_message = (
            None if is_complete else {"role": "assistant", "content": resp.content}
        )
        return ParsedTurn(
            text=text,
            tool_calls=tool_calls,
            is_complete=is_complete,
            assistant_message=assistant_message,
        )

    def _format_tool_result(
        self, call: ToolInvocation, content: str, is_error: bool
    ) -> anthropic.types.ToolResultBlockParam:
        result: anthropic.types.ToolResultBlockParam = {
            "type": "tool_result",
            "tool_use_id": call.id,
            "content": content,
        }
        if is_error:
            result["is_error"] = True
        return result

    def _append_tool_results(self, messages: list[Any], tool_results: list[Any]) -> None:
        messages.append({"role": "user", "content": tool_results})

    async def _create_with_retry(
        self,
        model: str,
        system_prompt: str,
        messages: list[anthropic.types.MessageParam],
        tools: list[anthropic.types.ToolParam],
    ) -> anthropic.types.Message:
        total = self.rate_limit_retries
        for attempt in range(1, total + 1):
            try:
                return await self.client.messages.create(
                    model=model,
                    max_tokens=16000,
                    system=[
                        {
                            "type": "text",
                            "text": system_prompt,
                            "cache_control": {"type": "ephemeral"},
                        }
                    ],
                    messages=messages,
                    tools=tools if tools else [],
                )
            except anthropic.RateLimitError:
                if attempt == total:
                    raise
                self._logger.warning(
                    f"Rate limit (429) on attempt {attempt}/{total} — sleeping {self.rate_limit_delay}s"
                )
                await asyncio.sleep(self.rate_limit_delay)
